src/pcre2_compile.c

```diff
@@ -6169,21 +6176,163 @@ for (;; pptr++)
 
     if ((*pptr & CLASS_IS_ECLASS) != 0)
       {
+      eclass_op_info op_info;
+      PCRE2_SIZE previous_length = (lengthptr != NULL)? *lengthptr : 0;
+      int classones = 0;
+      int classzeros = 0;
+
       previous = code;
       *code++ = OP_ECLASS;
       code += LINK_SIZE;
-      if (!PRIV(compile_class_nested)(options, xoptions, &pptr, &code,
-                                      errorcodeptr, cb, lengthptr))
+      *code++ = 0;  /* Flags, currently zero. */
+      if (!PRIV(compile_class_nested)(options, xoptions, FALSE, &pptr, &code,
+                                      &op_info, errorcodeptr, cb, lengthptr))
         return 0;
-      PUT(previous, 1, (int)(code - previous));
 
-      zeroreqcu = reqcu;
-      zeroreqcuflags = reqcuflags;
-      if (firstcuflags == REQ_UNSET) firstcuflags = REQ_NONE;
-      zerofirstcu = firstcu;
-      zerofirstcuflags = firstcuflags;
+      if (lengthptr != NULL)
+        {
+        *lengthptr += code - previous;
+        code = previous;
+        /* (*lengthptr - previous_length) now holds the amount of buffer that
+        we require to make the call to compile_class_nested() with
+        lengthptr = NULL, and including the (1+LINK_SIZE+1) that we write out
+        before that call. */
+        }
 
-      break;   /* We are finished with this class */
+      /* Do some useful counting of what's in the bitmap. */
+      while (classones < 32 && op_info.bits[classones] == 0xff) classones++;
+      while (classzeros < 32 && op_info.bits[classzeros] == 0) classzeros++;
+
+      /* After constant-folding the extended class syntax, it may turn out to be
+      a simple class after all. In that case, we can unwrap it from the
+      OP_ECLASS container - and in fact, we must do so, because in 8-bit
+      no-Unicode mode the matcher is compiled without support for OP_ECLASS. */
+
+#ifndef SUPPORT_WIDE_CHARS
+      PCRE2_ASSERT(op_info.op_single_type != 0);
+#else
+      if (op_info.op_single_type != 0)
+#endif
+        {
+        /* Rewind back over the OP_ECLASS. */
+        code = previous;
```

Comment:
Can we compute this at length computation time, store the bit in the META code, and use this knowledge when the byte code is generated. We might do a buffer overrun otherwise.

Reply:
There shouldn't be a buffer overrun?

This is the rule I'm following: "don't do any conditional logic / branches inside "`if (lengthptr == NULL)`" blocks". It's totally OK to check whether lengthptr == NULL, but you need to absolutely sure when you do that, that the code will behave the same with and without lengthptr.

What would be terrible, would be to do "better" optimisation when `lengthptr == NULL`, or to do anything different at all (apart from actually writing the bytecode itself). We have to make exactly the same optimisation decisions in each case. The `if (lengthptr == NULL)` blocks need to extremely short, and careful, so that they don't affect control flow.

That way, we can be sure we'll have enough buffer space - because we followed all the same code paths each time.

Reply:
Whenever I rewind the `code` pointer, I increment the `lengthptr` by the same amount.

* Say, we write two bytes to `code`, and request another 5 via `lengthptr`. (A total of 7 bytes requested)
* Then we rewind `code` (by 2 bytes), and increment `lengthptr` to compensate. (Still 7 bytes requested)
* Now we write more bytes to `code` - can initially decrement `lengthptr` (if writing 7 bytes or less), but then if we are requesting more bytes, we'll increase the size of the allocation